### PR TITLE
[misk.client] Records metrics for timeout requests

### PR DESCRIPTION
**Pull request by frojasg**

### CODE REVIEW

File: misk/src/test/kotlin/misk/client/ClientMetricsInterceptorTest.kt [hunk 1]
```diff
@@ -1,123 +1,117 @@
...
-            "Content-type: " + MediaTypes.APPLICATION_JSON)
-    fun ping(@Body request: AppRequest): Call<AppResponse>
-  }
+  @Test
+  fun timeouts() {
+    assertThatExceptionOfType(ExecutionException::class.java).isThrownBy {
+      client.ping(AppRequest(200)).execute().code()
+    }.withCauseInstanceOf(SocketTimeoutException::class.java)
 
-  class PingAction @Inject constructor() : WebAction {
```
Comment (frojasg) [inline misk/src/test/kotlin/misk/client/ClientMetricsInterceptorTest.kt]: It was easier to simulate timeouts by using  okhttp `MockWebServer`.

File: misk/src/test/kotlin/misk/client/ClientMetricsInterceptorTest.kt [hunk 2]
```diff
@@ -1,123 +1,117 @@
...
     @Provides
     @Singleton
-    fun provideHttpClientConfig(): HttpClientsConfig {
+    fun provideHttpClientConfig(server: MockWebServer): HttpClientsConfig {
+      val url = server.url("/")
       return HttpClientsConfig(
-              endpoints = mapOf("pinger" to HttpClientEndpointConfig(jetty.httpServerUrl.toString())))
+          endpoints = mapOf("pinger" to HttpClientEndpointConfig(
+              url = url.toString(),
+              readTimeout = Duration.ofMillis(10)
```
Comment (frojasg) [inline misk/src/test/kotlin/misk/client/ClientMetricsInterceptorTest.kt]: short timeout so we don't slow down the tests

File: misk/src/main/kotlin/misk/client/ClientMetricsInterceptor.kt [hunk 3]
```diff
@@ -13,14 +16,20 @@ class ClientMetricsInterceptor internal constructor(
...
+      val result = chain.proceed(chain.request)
+      val elapsedMillis = stopwatch.stop().elapsed(TimeUnit.MILLISECONDS).toDouble()
+      requestDuration.record(elapsedMillis, actionName, "all")
+      requestDuration.record(elapsedMillis, actionName, "${result.code / 100}xx")
+      requestDuration.record(elapsedMillis, actionName, "${result.code}")
+      return result
+    } catch (e: SocketTimeoutException) {
+      val elapsedMillis = stopwatch.stop().elapsed(TimeUnit.MILLISECONDS).toDouble()
+      requestDuration.record(elapsedMillis, actionName, "all")
+      requestDuration.record(elapsedMillis, actionName, "timeout")
```
Comment (wesleyk) [inline misk/src/main/kotlin/misk/client/ClientMetricsInterceptor.kt]: is this that useful? It's essentially what the client timeout is set to, no?

File: misk/src/test/kotlin/misk/client/ClientMetricsInterceptorTest.kt [hunk 4]
```diff
@@ -1,123 +1,117 @@
...
-    @POST("/ping")
-    @Headers(
-            "Accept: " + MediaTypes.APPLICATION_JSON,
-            "Content-type: " + MediaTypes.APPLICATION_JSON)
-    fun ping(@Body request: AppRequest): Call<AppResponse>
-  }
+  @Test
+  fun timeouts() {
+    assertThatExceptionOfType(ExecutionException::class.java).isThrownBy {
+      client.ping(AppRequest(200)).execute().code()
```
Comment (wesleyk) [inline misk/src/test/kotlin/misk/client/ClientMetricsInterceptorTest.kt]: interesting, where do we setup the mock server to timeout?

File: misk/src/main/kotlin/misk/client/ClientMetricsInterceptor.kt (hunk 3 above)
Comment (frojasg) [reply misk/src/main/kotlin/misk/client/ClientMetricsInterceptor.kt]: Yeah, the histogram metric is not very interesting other than validate you're timing out on the expected time. I was more interested to know how many times outs the client is seeing. Also, it consistent with the other metrics. Let me know if you want to use other metric.

File: misk/src/test/kotlin/misk/client/ClientMetricsInterceptorTest.kt (hunk 4 above)
Comment (frojasg) [reply misk/src/test/kotlin/misk/client/ClientMetricsInterceptorTest.kt]: You don't. The mock server only returns the enqueued requests. If you don't queue anything, it won't return, causing a timeout.

File: misk/src/main/kotlin/misk/client/ClientMetricsInterceptor.kt (hunk 3 above)
Comment (swankjesse) [reply misk/src/main/kotlin/misk/client/ClientMetricsInterceptor.kt]: I think it’s useful in that if it spikes it tells us something

**Pull request merged**